st2common/st2common/models/db/__init__.py

```diff
@@ -292,23 +296,34 @@ def get_all(self, *args, **kwargs):
         return self.query(*args, **kwargs)
 
     def count(self, *args, **kwargs):
-        result = self.model.objects(**kwargs).count()
+        result = self.model.objects(*args, **kwargs).count()
         log_query_and_profile_data_for_queryset(queryset=result)
         return result
 
-    def query(self, offset=0, limit=None, order_by=None, exclude_fields=None,
-              **filters):
+    # TODO: PEP-3102 introduced keyword-only arguments, so once we support Python 3+, we can change
```

Comment:
I would prefer to change pep8 rule or whatever the rule existing code was violating and keep it as it was.

And yeah correct long term solution is to get rid of `**filters` and replace it with explicit argument definition.

Reply:
Unfortunately that's not possible AFAIK. The syntax for keyword-only arguments in Python 3 causes a `SyntaxError` in Python 2.7. This is one of the changes in Python 3 that isn't backwards compatible.

It is, however, a good reason to start coding in Python 3. 😜

Reply:
Is it possible to always use keyword arguments when calling function then? I assume that's what we do right now and that's why we don't need `*args*.

Reply:
It's possible to force all arguments to be keyword-only, but I would recommend against that.

Right now the entire `MongoDBAccess` class is simply a wrapper around `mongoengine`, and `MongoDBAccess` tweaks filter arguments a bit to allow more expressiveness in queries. The `mongoengine` module is explicitly trying to copy Django's ORM API:

> It uses a simple declarative API, similar to the Django ORM.

I've developed a few websites with Django, which is where [`Q` objects came from](https://docs.djangoproject.com/en/1.11/topics/db/queries/#complex-lookups-with-q-objects).

Because our `MongoDBAccess` class is slightly closer to Django's ORM (we may want to look into upstreaming the `_process_arg_filters`, `_process_datetime_range_filters`, and `_process_null_filters` to `mongoengine`), it's (hopefully/presumably) easier for Python webdevs to contribute changes to StackStorm.

So I would keep `Q` objects as positional arguments and force pretty much every other argument into being keyword-only, simply because that's how Django and `mongoengine` do it.

And you are correct - (as far as I found) every use of `query` and `get` passes the `offset`, `limit`, `order_by`, and `exclude_fields` parameters as keyword arguments. However, passing them as keyword arguments hasn't been previously enforced. This change forces them to be.